File: lib/python/opentoken-cli/src/main/opentoken_cli/commands/package_command.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import logging
from typing import List

from opentoken.metadata import Metadata
from opentoken.tokentransformer.encrypt_token_transformer import EncryptTokenTransformer
from opentoken.tokentransformer.hash_token_transformer import HashTokenTransformer
from opentoken.tokentransformer.token_transformer import TokenTransformer
from opentoken_cli.io.csv.person_attributes_csv_reader import PersonAttributesCSVReader
from opentoken_cli.io.csv.person_attributes_csv_writer import PersonAttributesCSVWriter
from opentoken_cli.io.json.metadata_json_writer import MetadataJsonWriter
from opentoken_cli.io.parquet.person_attributes_parquet_reader import PersonAttributesParquetReader
from opentoken_cli.io.parquet.person_attributes_parquet_writer import PersonAttributesParquetWriter
from opentoken_cli.processor.person_attributes_processor import PersonAttributesProcessor

logger = logging.getLogger(__name__)


class PackageCommand:
    """
    Package command - combines tokenize and encrypt in one command.
    This is the default workflow: hash + encrypt.
    """

    TYPE_CSV = "csv"
    TYPE_PARQUET = "parquet"

    @staticmethod
    def register_subcommand(subparsers):
        """Register the package subcommand with the argument parser."""
        parser = subparsers.add_parser(
            "package",
            help="Generate and encrypt tokens in one step (tokenize + encrypt)",
            description="Generate and encrypt tokens in one step",
        )

        parser.add_argument(
            "-i",
            "--input",
            required=True,
            dest="input_path",
            help="Input file path",
        )

        parser.add_argument(
            "-o",
            "--output",
            required=True,
            dest="output_path",
            help="Output file path",
        )

        parser.add_argument(
            "-t",
            "--input-type",
            required=True,
            dest="input_type",
            choices=["csv", "parquet"],
            help="Input file type: csv or parquet",
        )

        parser.add_argument(
            "-ot",
            "--output-type",
            dest="output_type",
            choices=["csv", "parquet"],
            help="Output file type (defaults to input type): csv or parquet",
        )

        parser.add_argument(
            "--hashingsecret",
            required=True,
            dest="hashing_secret",
            help="Hashing secret for token generation",
        )

        parser.add_argument(
            "--encryptionkey",
            required=True,
            dest="encryption_key",
            help="Encryption key for token encryption",
        )

        parser.set_defaults(func=PackageCommand.execute)

    @staticmethod
    def execute(args):
        """Execute the package command."""
        logger.info("Running package command (tokenize + encrypt)")

        # Default output type to input type if not specified
        output_type = args.output_type if args.output_type else args.input_type

        # Log parameters (mask secrets)
        logger.info(f"Input: {args.input_path} ({args.input_type})")
        logger.info(f"Output: {args.output_path} ({output_type})")
        logger.info(f"Hashing Secret: {PackageCommand._mask_string(args.hashing_secret)}")
        logger.info(f"Encryption Key: {PackageCommand._mask_string(args.encryption_key)}")

        # Validate secrets
        if not args.hashing_secret or not args.hashing_secret.strip():
            logger.error("Hashing secret is required")
            return 1
        if not args.encryption_key or not args.encryption_key.strip():
            logger.error("Encryption key is required")
            return 1

        try:
            PackageCommand._process_tokens(
                args.input_path,
                args.output_path,
                args.input_type,
                output_type,
                args.hashing_secret,
                args.encryption_key,
            )
            logger.info("Token generation and encryption completed successfully")
            return 0
        except Exception as e:
            logger.error(f"Error during token processing: {e}", exc_info=True)
            return 1

    @staticmethod
    def _process_tokens(
        input_path: str,
        output_path: str,
        input_type: str,
        output_type: str,
        hashing_secret: str,
        encryption_key: str,
    ):
        """Process tokens from person attributes."""
        token_transformer_list: List[TokenTransformer] = []

        try:
            # Add both hash and encryption transformers
            token_transformer_list.append(HashTokenTransformer(hashing_secret))
            token_transformer_list.append(EncryptTokenTransformer(encryption_key))
        except Exception as e:
            logger.error("Error initializing transformers", exc_info=e)
            raise RuntimeError("Failed to initialize transformers") from e

        try:
            with PackageCommand._create_reader(
                input_path, input_type
            ) as reader, PackageCommand._create_writer(output_path, output_type) as writer:

                # Create metadata
                metadata = Metadata()
                metadata_map = metadata.initialize()
                metadata.add_hashed_secret(Metadata.HASHING_SECRET_HASH, hashing_secret)
                metadata.add_hashed_secret(Metadata.ENCRYPTION_SECRET_HASH, encryption_key)

                # Process data
                PersonAttributesProcessor.process(reader, writer, token_transformer_list, metadata_map)

                # Write metadata
                metadata_writer = MetadataJsonWriter(output_path)
                metadata_writer.write(metadata_map)

        except Exception as e:
            logger.error("Error processing tokens", exc_info=e)
            raise

    @staticmethod
    def _create_reader(path: str, file_type: str):
        """Create a PersonAttributesReader based on file type."""
        file_type_lower = file_type.lower()
        if file_type_lower == PackageCommand.TYPE_CSV:
            return PersonAttributesCSVReader(path)
        elif file_type_lower == PackageCommand.TYPE_PARQUET:
            return PersonAttributesParquetReader(path)
        else:
            raise ValueError(f"Unsupported input type: {file_type}")

    @staticmethod
    def _create_writer(path: str, file_type: str):
        """Create a PersonAttributesWriter based on file type."""
        file_type_lower = file_type.lower()
        if file_type_lower == PackageCommand.TYPE_CSV:
            return PersonAttributesCSVWriter(path)
        elif file_type_lower == PackageCommand.TYPE_PARQUET:
            return PersonAttributesParquetWriter(path)
        else:
            raise ValueError(f"Unsupported output type: {file_type}")

    @staticmethod
    def _mask_string(input_str: str) -> str:
        """Mask a string for logging purposes, showing only first 3 characters."""
        if input_str is None or len(input_str) <= 3:
            return input_str
        return input_str[:3] + "*" * (len(input_str) - 3)

File: lib/java/opentoken-cli/src/main/java/com/truveta/opentoken/cli/commands/PackageCommand.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.cli.commands;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.truveta.opentoken.Metadata;
import com.truveta.opentoken.cli.io.MetadataWriter;
import com.truveta.opentoken.cli.io.PersonAttributesReader;
import com.truveta.opentoken.cli.io.PersonAttributesWriter;
import com.truveta.opentoken.cli.io.csv.PersonAttributesCSVReader;
import com.truveta.opentoken.cli.io.csv.PersonAttributesCSVWriter;
import com.truveta.opentoken.cli.io.json.MetadataJsonWriter;
import com.truveta.opentoken.cli.io.parquet.PersonAttributesParquetReader;
import com.truveta.opentoken.cli.io.parquet.PersonAttributesParquetWriter;
import com.truveta.opentoken.cli.processor.PersonAttributesProcessor;
import com.truveta.opentoken.tokentransformer.EncryptTokenTransformer;
import com.truveta.opentoken.tokentransformer.HashTokenTransformer;
import com.truveta.opentoken.tokentransformer.TokenTransformer;

import picocli.CommandLine.Command;
import picocli.CommandLine.Option;

/**
 * Package command - combines tokenize and encrypt in one command.
 * This is the default workflow: hash + encrypt.
 */
@Command(
    name = "package",
    description = "Generate and encrypt tokens in one step (tokenize + encrypt)",
    mixinStandardHelpOptions = true
)
public class PackageCommand implements Callable<Integer> {
    
    private static final Logger logger = LoggerFactory.getLogger(PackageCommand.class);
    private static final String TYPE_CSV = "csv";
    private static final String TYPE_PARQUET = "parquet";
    
    @Option(names = {"-i", "--input"}, required = true,
            description = "Input file path")
    private String inputPath;
    
    @Option(names = {"-o", "--output"}, required = true,
            description = "Output file path")
    private String outputPath;
    
    @Option(names = {"-t", "--input-type"}, required = true,
            description = "Input file type: csv or parquet")
    private String inputType;
    
    @Option(names = {"-ot", "--output-type"},
            description = "Output file type (defaults to input type): csv or parquet")
    private String outputType;
    
    @Option(names = {"-h", "--hashingsecret"}, required = true,
            description = "Hashing secret for token generation")
    private String hashingSecret;
    
    @Option(names = {"-e", "--encryptionkey"}, required = true,
            description = "Encryption key for token encryption")
    private String encryptionKey;
    
    @Override
    public Integer call() {
        logger.info("Running package command (tokenize + encrypt)");
        
        // Default output type to input type if not specified
        if (outputType == null || outputType.isEmpty()) {
            outputType = inputType;
        }
        
        // Log parameters (mask secrets)
        logger.info("Input: {} ({})", inputPath, inputType);
        logger.info("Output: {} ({})", outputPath, outputType);
        logger.info("Hashing Secret: {}", maskString(hashingSecret));
        logger.info("Encryption Key: {}", maskString(encryptionKey));
        
        // Validate types
        if (!isValidType(inputType)) {
            logger.error("Invalid input type: {}. Must be 'csv' or 'parquet'", inputType);
            return 1;
        }
        if (!isValidType(outputType)) {
            logger.error("Invalid output type: {}. Must be 'csv' or 'parquet'", outputType);
            return 1;
        }
        
        // Validate secrets
        if (hashingSecret == null || hashingSecret.isBlank()) {
            logger.error("Hashing secret is required");
            return 1;
        }
        if (encryptionKey == null || encryptionKey.isBlank()) {
            logger.error("Encryption key is required");
            return 1;
        }
        
        try {
            processTokens();
            logger.info("Token generation and encryption completed successfully");
            return 0;
        } catch (Exception e) {
            logger.error("Error during token processing", e);
            return 1;
        }
    }
    
    private void processTokens() throws IOException {
        List<TokenTransformer> transformers = new ArrayList<>();
        
        try {
            // Add both hash and encryption transformers
            transformers.add(new HashTokenTransformer(hashingSecret));
            transformers.add(new EncryptTokenTransformer(encryptionKey));
        } catch (Exception e) {
            logger.error("Error initializing transformers", e);
            throw new RuntimeException("Failed to initialize transformers", e);
        }
        
        try (PersonAttributesReader reader = createReader(inputPath, inputType);
             PersonAttributesWriter writer = createWriter(outputPath, outputType)) {
            
            // Create metadata
            Metadata metadata = new Metadata();
            Map<String, Object> metadataMap = metadata.initialize();
            metadata.addHashedSecret(Metadata.HASHING_SECRET_HASH, hashingSecret);
            metadata.addHashedSecret(Metadata.ENCRYPTION_SECRET_HASH, encryptionKey);
            
            // Process data
            PersonAttributesProcessor.process(reader, writer, transformers, metadataMap);
            
            // Write metadata
            MetadataWriter metadataWriter = new MetadataJsonWriter(outputPath);
            metadataWriter.write(metadataMap);
        } catch (Exception e) {
            logger.error("Error processing tokens", e);
            throw new IOException("Failed to process tokens", e);
        }
    }
    
    private PersonAttributesReader createReader(String path, String type) throws IOException {
        return switch (type.toLowerCase()) {
            case TYPE_CSV -> new PersonAttributesCSVReader(path);
            case TYPE_PARQUET -> new PersonAttributesParquetReader(path);
            default -> throw new IllegalArgumentException("Unsupported input type: " + type);
        };
    }
    
    private PersonAttributesWriter createWriter(String path, String type) throws IOException {
        return switch (type.toLowerCase()) {
            case TYPE_CSV -> new PersonAttributesCSVWriter(path);
            case TYPE_PARQUET -> new PersonAttributesParquetWriter(path);
            default -> throw new IllegalArgumentException("Unsupported output type: " + type);
        };
    }
    
    private boolean isValidType(String type) {
        return TYPE_CSV.equalsIgnoreCase(type) || TYPE_PARQUET.equalsIgnoreCase(type);
    }
    
    private String maskString(String input) {
        if (input == null || input.length() <= 3) {
            return input;
        }
        return input.substring(0, 3) + "*".repeat(input.length() - 3);
    }
}
